Opamp usability

KozhinovAlexander opened the pull request:
Followinmg changes were introduced:

[drivers: opamp: Kconfig: log module dependency](https://github.com/zephyrproject-rtos/zephyr/pull/98877/commits/2f7b551bd141c0341008082d5abd308402c0724b):
  use common log module dependency pattern in KConfig
  add LOG as module dependency for opamp driver

[dts: bindings: opamp: opamp-controller.yaml](https://github.com/zephyrproject-rtos/zephyr/pull/98877/commits/88802a0c16d945baee8188e5af2ac5dd20f4ca02):
  make functional mode required
  add functional modes description with circuit examples

[samples: drivers: opamp: output_measure: prj.conf](https://github.com/zephyrproject-rtos/zephyr/pull/98877/commits/90f3bf284b1f416179126ac79950e5ad2fcc9756):
  enable logging
  add log buffer size and debug options and comment them

[samples: drivers: opamp: output_measure](https://github.com/zephyrproject-rtos/zephyr/pull/98877/commits/3d940a05243d504c2053161fb9b8e5e4b459d1cb):
  show gain value at runtime as a starting
  use ADC calibration before conversion
  update README with sample gain prints

KozhinovAlexander committed:
dts: bindings: opamp: opamp-controller.yaml

make functional mode required
add functional modes description with circuit examples

Signed-off-by: Alexander Kozhinov <ak.alexander.kozhinov@gmail.com>

KozhinovAlexander committed:
samples: drivers: opamp: output_measure

prj.conf:
	enable logging
	add log buffer size and debug options and comment them

main.c:
	show gain value at runtime as a starting
	use ADC calibration before conversion

README.rst:
	update with sample gain prints

Signed-off-by: Alexander Kozhinov <ak.alexander.kozhinov@gmail.com>

KozhinovAlexander committed:
dts: arm: nxp: opamp functional mode

set the default functional mode of OPAMP peripherals to "follower"

Signed-off-by: Alexander Kozhinov <ak.alexander.kozhinov@gmail.com>

KozhinovAlexander commented on boards/nxp/frdm_mcxa156/frdm_mcxa156.dts:
```diff
@@ -239,6 +239,7 @@ zephyr_mipi_dbi_parallel: &flexio0_lcd {
 	status = "okay";
 	pinctrl-0 = <&pinmux_opamp0>;
 	pinctrl-names = "default";
+	functional-mode = "follower";
```
moved to soc level

ZhaoxiangJin reviewed (changes requested):
1. commits not bisectable
2. not clear why `module-dep` and `module-help` are added, since they are not used IMO.

ZhaoxiangJin reviewed (commented):
no more change request, I will do test the next day, thanks for your effort.

KozhinovAlexander commented:
> when i build and run the sample base on your branch, i got:
> 
> ```
> *** Booting ZOPAMP output (@gain: 4) is: 1556(mv)
> ephyr OS buiOPAMP output (@gain: 2) is: 937(mv)
> ld v4.3.0-rc2OPAMP output (@gain: 1) is: 623(mv)
> -101-g65eb32acd9bb ***
> ```
> 
> Could you please help replace all printf calls in main.c with printk?

Changed to printk and rebased on top of newest main.

KozhinovAlexander commented:
> > > when i build and run the sample base on your branch, i got:
> > > ```
> > > *** Booting ZOPAMP output (@gain: 4) is: 1556(mv)
> > > ephyr OS buiOPAMP output (@gain: 2) is: 937(mv)
> > > ld v4.3.0-rc2OPAMP output (@gain: 1) is: 623(mv)
> > > -101-g65eb32acd9bb ***
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Could you please help replace all printf calls in main.c with printk?
> > 
> > 
> > Changed to printk and rebased on top of newest main.
> 
> It would be better to change all printf to printk in the main.c, could you please help to update?

done. all printf were changed to printk

KozhinovAlexander commented:
@ZhaoxiangJin please take a look

kartben commented on drivers/opamp/Kconfig:
```diff
@@ -9,7 +10,7 @@ menuconfig OPAMP
 
 module = OPAMP
-module-str = opamp
+module-str = OPAMP driver
```
given the original version is what 99.5% of the drivers use as a convention, I would not change it

KozhinovAlexander replied:
Sure, I'll do. Thank you both.

KozhinovAlexander replied:
done

KozhinovAlexander commented on dts/bindings/opamp/opamp-controller.yaml:
```diff
@@ -17,6 +18,94 @@ properties:
+               Rin        Rf      |
+                                 GND
+
+      e) standalone Mode
```
@copilot open a new pull request to apply changes based on [this feedback](https://github.com/zephyrproject-rtos/zephyr/pull/98877#discussion_r2511908404)

kartben replied:
I doubt this will work (especially w.r.t authorship/ sign off) but would be cool if it does :)

etienne-lms commented on dts/bindings/opamp/opamp-controller.yaml:
```diff
@@ -17,6 +18,93 @@ properties:
     description: |
       Selects opamp functional mode.
 
+      a) inverting mode
```
Maybe order in the same order used in the enumerated list above.

KozhinovAlexander replied:
Thnx. Done

etienne-lms commented on samples/drivers/opamp/output_measure/src/main.c:
```diff
@@ -97,32 +119,36 @@ int main(void)
 		return 0;
 	}
 
+	enum opamp_gain gain_value = (enum opamp_gain)(OPAMP_GAIN_64 + 1); /* default gain */
```
This makes `enum opamp_gain` a value that is not part of the enumerated list. Some toolchain may complain.
Maybe consider:
```patch
+	const char *gain_str = "default";

	do {
#if OPAMP_SUPPORT_PROGRAMMABLE_GAIN
		/* For OPAMPs which support programmable gain, retrieve the gain array. */
		ret = opamp_set_gain(opamp_dev, gain[size - 1]);		
		if (ret < 0) {
			printk("OPAMP gain set failed with code (%d)\n", ret);
			return 0;
		}

+		gain_str = opamp_gain2str[gain[size - 1]];
		size--;
#endif
		ret = read_opamp_output(&adc_channel, &sequence, &sample_buffer);
		if (ret < 0) {
			printk("OPAMP output read failed with code (%d)\n", ret);
			return 0;
		}

-		printf("OPAMP output is: %d(mv)\n", sample_buffer);
+		printk("OPAMP output (@gain: %s) is: %d(mv)\n", gain_str, sample_buffer);

etienne-lms reviewed (commented):
For commit "samples: drivers: opamp: output_measure".
Could you mention in the commit message that `printf()` is replaced with `printk()` (and preferably a word on the rational).
The commit message header line misses something, like:
```patch
-samples: drivers: opamp: output_measure
+samples: drivers: opamp: output_measure: enable calibration and log gain

KozhinovAlexander commented on samples/drivers/opamp/output_measure/src/main.c:
```diff
@@ -97,32 +118,34 @@ int main(void)
 		return 0;
 	}
 
+	const char *gain_str = "default";
```
I see it as a cosmetic change, because style-checker does not throw any error. We could really make our reviews faster, if we avoid cosmetic changes 😉
UPD: I will change.

KozhinovAlexander replied:
done

KozhinovAlexander commented:
@etienne-lms Pleae take a look at your change-requests and resolve if they're modified correctly.

etienne-lms reviewed (approved):
Non-blocking comment.

KozhinovAlexander commented on samples/drivers/opamp/output_measure/src/main.c:
```diff
@@ -24,31 +24,52 @@ uint8_t size = ARRAY_SIZE(gain);
 uint8_t size;
 #endif
 
+const static char *opamp_gain2str[] = {
```
Therefore I would prefer to drop this one.
